File: wechatpay-medical-insurance-payment/references/2-服务商/示例代码/Go/2-订单查询/query_by_mix_trade_no.go
package main

import (
	"demo/wxpay_utility"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

func main() {
	config, err := wxpay_utility.CreateMchConfig(
		"1900000100",
		"1DDE55AD98Exxxxxxxxxx",
		"/path/to/apiclient_key.pem",
		"PUB_KEY_ID_xxxxxxxxxxxxx",
		"/path/to/wxp_pub.pem",
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &QueryRequest{
		MixTradeNo: wxpay_utility.String("202204022005169952975171534816"),
		SubMchid:   wxpay_utility.String("1900000109"),
	}

	response, err := QueryByMixTradeNo(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		return
	}
	fmt.Printf("请求成功: %+v\n", response)
}

func QueryByMixTradeNo(config *wxpay_utility.MchConfig, request *QueryRequest) (response *OrderEntity, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "GET"
		path   = "/v3/med-ins/orders/mix-trade-no/{mix_trade_no}"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqUrl.Path = strings.Replace(reqUrl.Path, "{mix_trade_no}", url.PathEscape(*request.MixTradeNo), -1)
	query := reqUrl.Query()
	if request.SubMchid != nil {
		query.Add("sub_mchid", *request.SubMchid)
	}
	reqUrl.RawQuery = query.Encode()

	httpRequest, err := http.NewRequest(method, reqUrl.String(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		err = wxpay_utility.ValidateResponse(config.WechatPayPublicKeyId(), config.WechatPayPublicKey(), &httpResponse.Header, respBody)
		if err != nil {
			return nil, err
		}
		response := &OrderEntity{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}
		return response, nil
	}
	return nil, wxpay_utility.NewApiException(httpResponse.StatusCode, httpResponse.Header, respBody)
}

type QueryRequest struct {
	MixTradeNo *string `json:"mix_trade_no,omitempty"`
	SubMchid   *string `json:"sub_mchid,omitempty"`
}

func (o *QueryRequest) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct{}{})
}

type OrderEntity struct {
	MixTradeNo       *string `json:"mix_trade_no,omitempty"`
	MixPayStatus     *string `json:"mix_pay_status,omitempty"`
	SelfPayStatus    *string `json:"self_pay_status,omitempty"`
	MedInsPayStatus  *string `json:"med_ins_pay_status,omitempty"`
	PaidTime         *string `json:"paid_time,omitempty"`
	MixPayType       *string `json:"mix_pay_type,omitempty"`
	OrderType        *string `json:"order_type,omitempty"`
	Appid            *string `json:"appid,omitempty"`
	SubAppid         *string `json:"sub_appid,omitempty"`
	SubMchid         *string `json:"sub_mchid,omitempty"`
	Openid           *string `json:"openid,omitempty"`
	SubOpenid        *string `json:"sub_openid,omitempty"`
	OutTradeNo       *string `json:"out_trade_no,omitempty"`
	TotalFee         *int64  `json:"total_fee,omitempty"`
	MedInsGovFee     *int64  `json:"med_ins_gov_fee,omitempty"`
	MedInsSelfFee    *int64  `json:"med_ins_self_fee,omitempty"`
	MedInsOtherFee   *int64  `json:"med_ins_other_fee,omitempty"`
	MedInsCashFee    *int64  `json:"med_ins_cash_fee,omitempty"`
	WechatPayCashFee *int64  `json:"wechat_pay_cash_fee,omitempty"`
	MedInsFailReason *string `json:"med_ins_fail_reason,omitempty"`
}

File: wechatpay-medical-insurance-payment/references/1-商户/示例代码/Java/2-订单查询/QueryByMixTradeNo.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/merchant/4014931831

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.List;

/**
 * 使用医保自费混合订单号查看下单结果
 */
public class QueryByMixTradeNo {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "GET";
  private static String PATH = "/v3/med-ins/orders/mix-trade-no/{mix_trade_no}";

  public static void main(String[] args) {
    QueryByMixTradeNo client = new QueryByMixTradeNo(
      "19xxxxxxxx",
      "1DDE55AD98Exxxxxxxxxx",
      "/path/to/apiclient_key.pem",
      "PUB_KEY_ID_xxxxxxxxxxxxx",
      "/path/to/wxp_pub.pem"
    );

    QueryRequest request = new QueryRequest();
    request.mixTradeNo = "202204022005169952975171534816";
    try {
      OrderEntity response = client.run(request);
      System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
      e.printStackTrace();
    }
  }

  public OrderEntity run(QueryRequest request) {
    String uri = PATH.replace("{mix_trade_no}", WXPayUtility.urlEncode(request.mixTradeNo));

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, null));
    reqBuilder.method(METHOD, null);
    Request httpRequest = reqBuilder.build();

    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey, httpResponse.headers(), respBody);
        return WXPayUtility.fromJson(respBody, OrderEntity.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public QueryByMixTradeNo(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class QueryRequest {
    @SerializedName("mix_trade_no")
    @Expose(serialize = false)
    public String mixTradeNo;
  }

  /** 应答结构与下单接口一致，新增 med_ins_fail_reason（失败原因，仅查询时返回） */
  public static class OrderEntity {
    @SerializedName("mix_trade_no") public String mixTradeNo;
    @SerializedName("mix_pay_status") public String mixPayStatus;
    @SerializedName("self_pay_status") public String selfPayStatus;
    @SerializedName("med_ins_pay_status") public String medInsPayStatus;
    @SerializedName("paid_time") public String paidTime;
    @SerializedName("passthrough_response_content") public String passthroughResponseContent;
    @SerializedName("mix_pay_type") public String mixPayType;
    @SerializedName("order_type") public String orderType;
    @SerializedName("appid") public String appid;
    @SerializedName("openid") public String openid;
    @SerializedName("pay_for_relatives") public Boolean payForRelatives;
    @SerializedName("out_trade_no") public String outTradeNo;
    @SerializedName("serial_no") public String serialNo;
    @SerializedName("pay_order_id") public String payOrderId;
    @SerializedName("pay_auth_no") public String payAuthNo;
    @SerializedName("geo_location") public String geoLocation;
    @SerializedName("city_id") public String cityId;
    @SerializedName("med_inst_name") public String medInstName;
    @SerializedName("med_inst_no") public String medInstNo;
    @SerializedName("med_ins_order_create_time") public String medInsOrderCreateTime;
    @SerializedName("total_fee") public Long totalFee;
    @SerializedName("med_ins_gov_fee") public Long medInsGovFee;
    @SerializedName("med_ins_self_fee") public Long medInsSelfFee;
    @SerializedName("med_ins_other_fee") public Long medInsOtherFee;
    @SerializedName("med_ins_cash_fee") public Long medInsCashFee;
    @SerializedName("wechat_pay_cash_fee") public Long wechatPayCashFee;
    @SerializedName("callback_url") public String callbackUrl;
    @SerializedName("prepay_id") public String prepayId;
    @SerializedName("attach") public String attach;
    @SerializedName("channel_no") public String channelNo;
    @SerializedName("med_ins_test_env") public Boolean medInsTestEnv;
    @SerializedName("med_ins_fail_reason") public String medInsFailReason;
  }
}
